pkgs/top-level/all-packages.nix

```diff
@@ -15898,19 +15891,14 @@ in
   linuxPackages = linuxPackages_4_19;
   linux = linuxPackages.kernel;
 
-  # Update this when adding the newest kernel major version!
-  linuxPackages_latest = linuxPackages_5_3;
-  linux_latest = linuxPackages_latest.kernel;
-
   # Build the kernel modules for the some of the kernels.
+  linuxPackages_latest = linuxPackagesFor pkgs.linux_latest;
   linuxPackages_mptcp = linuxPackagesFor pkgs.linux_mptcp;
   linuxPackages_rpi = linuxPackagesFor pkgs.linux_rpi;
   linuxPackages_4_4 = recurseIntoAttrs (linuxPackagesFor pkgs.linux_4_4);
   linuxPackages_4_9 = recurseIntoAttrs (linuxPackagesFor pkgs.linux_4_9);
   linuxPackages_4_14 = recurseIntoAttrs (linuxPackagesFor pkgs.linux_4_14);
   linuxPackages_4_19 = recurseIntoAttrs (linuxPackagesFor pkgs.linux_4_19);
-  linuxPackages_5_2 = recurseIntoAttrs (linuxPackagesFor pkgs.linux_5_2);
-  linuxPackages_5_3 = recurseIntoAttrs (linuxPackagesFor pkgs.linux_5_3);
```

Comment:
I think we keep the versioned name for "latest"; @NeQuissimus backports the changes from master, and `_latest` always points to a versioned attribute, which gets dropped as needed.

Reply:
Wouldn’t that mean we end up with unstable attribute names, which is what that part of the manual is trying to avoid?

Reply:
It's much easier if `latest` is just a "symlink" to a named attribute.
We could drop 5.2, I suppose... I generally keep the same kernels in stable that we have in master, except I don't add new ones unless requested. So unsupported ones go away, no new ones are added. Although I will likely bring 5.4 to 19.09 since it is probably going to be the next LTS kernel.

Reply:
Just verifying, [all updates should be backported](https://github.com/NixOS/nixpkgs/issues/50412#issuecomment-439258275), considering it is sometimes necessary for new hardware to work.

Anyway, yeah, this can break the workflow of someone that specifically used a versioned attribute, but this is not an issue per se. It would be an issue if there is no upgrade path to the next stable version.

Reply:
Ah yes, we did discuss that :) :+1:

Reply:
Hmm. This doesn’t match at all with [this discussion](https://logs.nix.samueldr.com/nixos-dev/2019-06-24#2333952;) that prompted this PR, nor with what the manual says (added in https://github.com/NixOS/nixpkgs/pull/63735).

Mentioning some people from that discussion: @domenkozar @Ekleog @grahamc @LnL7 @timokau

Reply:
Hm, I previously had a discussion on the same topic, with the conclusion that backporting the latest kernels, thus ensuring dropping EOL'd kernels won't revert `_latest` to a prior version was the way to go. That was before pinging @NeQuissimus on the topic.

I think we may need to revise the phrasing to drop the kernel example since we have a (de facto, I see) process to handle it.

Additionally, writing down the fact that the versioned kernel attributes and kernel upgrades are to be backported would be needed, I think.

And furthermore, we may want to codify the fact that the versioned kernel attributes will be dropped on EOL, both from stable and unstable when it happens.

Thoughts?

Reply:
I don’t think we should be providing attributes in stable that we don’t intend to exist for the lifetime of stable, and I’m surprised to learn that we are. Surely that’s the opposite of stability?

The reason the kernel is the example in the manual is because the kernel was the subject of the discussion we had that led to that manual section.

Reply:
(Sorry for dumping a wall of text.)

I think it depends on what is defined as stable. Is it a stable API/ABI for Nixpkgs? Is it stability of [the release] of the software included? I don't think it's an "either/or" question. And I'm likely missing some kind of stabilities.

Since stable does not mean stale, we're handling "major upgrade" backports to stable of some subsystems from unstable. Among those there are two main kinds, security sensitive software (e.g. browsers), and updates "to make stuff work" (e.g. spotify, youtube-dl) (though those examples are more service oriented).

New kernel versions are of the "to make stuff work" kind. This is because it's been proven in the past that new stuff sometimes doesn't work right with LTS, especially when it "skips" a NixOS upgrade.

The backports process is to cherry-pick changes from unstable, generally without changes. This is the main reason I see that the process of backporting kernel upgrades and EOL removal is fine as it is.

The previous issue where I approached @NeQuissimus was to codify the approach that was observed with the previous release (18.03), and make sure that no kernel versioning regression happened. I wasn't involved with NixOS with 17.09 and prior to know whether it was the practice at that time to backport upgrades to stable that way.

Though, with that said, I do appreciate the concerns on the issue of stable ABI, and disappearing attributes. It does make it harder to make a judgement call here on the issue. And with that said, pre-emptively dropping attribute names does not in itself break the codified process of bringing kernel updates to the stable branch.

This is only the thought process I have had up to here. I don't have an answer or judgement or anything more concrete. I only know what has happened for the other releases up to now.

* * *

Partly unrelated, and likely not to be an issue, but what about the introduction of a new LTS? Should it skip having its versioned attribute added to all-packages, and only be the `_latest` in a transient manner until next LTS+1 is released? I figure here that the answer is simple "LTS gets their versioned attribute added". It's only tangentially related to the issue at hand, but something to keep in mind. I'm not sure if this muddies the waters.

Reply:
> New kernel versions are of the "to make stuff work" kind. This is because it's been proven in the past that new stuff sometimes doesn't work right with LTS, especially when it "skips" a NixOS upgrade.

That is what users could use the `_latest` kernel for. With that attribute we can provide access to the latest kernel for the latest hardware, without guaranteeing/suggesting support for any particular non-lts kernel version. If users just pin at some version that works now and then forget about it, that may lead to security issues that won't be fixed.

Reply:
> And with that said, pre-emptively dropping attribute names does not in itself break the codified process of bringing kernel updates to the stable branch.

Yeah, exactly!  I’m all for bringing kernel updates to stable.

> what about the introduction of a new LTS? Should it skip having its versioned attribute added to all-packages, and only be the `_latest` in a transient manner until next LTS+1 is released? I figure here that the answer is simple "LTS gets their versioned attribute added".

Yes. I’d propose the following:

- When a non-LTS kernel is released, linux_latest is updated to be that kernel, but no attribute for that kernel is added.
- When an LTS kernel is released, an attribute is added, and linux_latest is updated to be an alias to that attribute.

That way, we end up with kernel upgrades in stable, which I think we all agree is a good thing, but we don’t end up with people writing configurations that break when they update their stable channel. I think that would also be a very good thing.